async-nats/src/jetstream/pull.rs

```diff
@@ -0,0 +1,192 @@
+
+        Ok(BatchStream {
+            pending_messages: batch.batch,
+            subscriber: consumer.subscription.as_mut().unwrap(),
+            context: consumer.context.clone(),
+        })
+    }
+}
+
+impl<'a> Stream for BatchStream<'a> {
+    type Item = Result<JetStreamMessage, Error>;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        if self.pending_messages == 0 {
+            return std::task::Poll::Ready(None);
+        }
+        match self.subscriber.receiver.poll_recv(cx) {
+            std::task::Poll::Ready(message) => {
+                if let Some(message) = message {
+                    if message.status == Some(StatusCode::TIMEOUT) {
+                        return Poll::Ready(None);
+                    }
+                    if message.status == Some(StatusCode::IDLE_HEARBEAT) {
+                        return std::task::Poll::Pending;
+                    }
+                    self.pending_messages -= 1;
+                    Poll::Ready(Some(Ok(JetStreamMessage {
+                        context: self.context.clone(),
+                        message,
+                    })))
+                } else {
+                    Poll::Ready(None)
+                }
+            }
+            std::task::Poll::Pending => std::task::Poll::Pending,
+        }
+    }
+}
+
+pub struct PullStream<'a> {
+    consumer: &'a mut Consumer<PullConsumerConfig>,
+    batch: usize,
+}
+
+impl<'a> PullStream<'a> {
+    fn new(consumer: &'a mut Consumer<PullConsumerConfig>, batch: usize) -> PullStream<'a> {
+        PullStream { consumer, batch }
+    }
+
+    pub fn process(&'a mut self) -> impl Stream<Item = Result<JetStreamMessage, Error>> + 'a {
+        try_stream! {
+            loop {
+                let  stream = self.consumer.batch(self.batch, None).await?;
+                for await message in stream {
+                    yield message?;
+                }
+            }
+        }
+    }
+}
+
+/// Used for next Pull Request for Pull Consumer
+#[derive(Debug, Default, Serialize, Deserialize, Clone, Copy, PartialEq, Eq)]
+pub struct BatchOptions {
+    /// The number of messages that are being requested to be delivered.
+    pub batch: usize,
+    /// The optional number of nanoseconds that the server will store this next request for
+    /// before forgetting about the pending batch size.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub expires: Option<usize>,
+    /// This optionally causes the server not to store this pending request at all, but when there are no
+    /// messages to deliver will send a nil bytes message with a Status header of 404, this way you
+    /// can know when you reached the end of the stream for example. A 409 is returned if the
+    /// Consumer has reached MaxAckPending limits.
+    #[serde(default, skip_serializing_if = "is_default")]
+    pub no_wait: bool,
```

Comment:
Variants that use skip serialize, would probably be better off as overloads.

```suggestion
    pub no_wait: bool,
```

Reply:
I think this will become unnecesarily messy. Skipping serialization is basically a way of saving a few bytes over the wire, nothing else.